Get rid of gcc-cross-wrapper

Ericson2314 opened the pull request:
###### Motivation for this change

We want to get rid of the old gcc-cross-wrapper cause it is old, jank, and redundant. Also, it won't generalize to non-gcc for e.g cross compiling on Darwin. This PR just worries about linux, but ensuring no regressions there is work enough. Darwin support will be a done in a follow-up PR (probably repurposed #25047).

This looks stupendously big, but only the first few commits (already split for easier reviewing) actually change infrastructure. The rest are rote fix-ups to get tests to pass that do not need to be reviewed so closely.

###### Things done

http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-scratch 2/3 of the PR is cleaning up derivations so those tests don't regress!

No native hashes should be changed by this commit

---

Huge shout-out to @Dridus for helping author this massive changeset!

CC @matthewbauer

Ericson2314 committed:
cc-wrapper: Learn about target prefixes

This is first step towards getting rid of gcc-wrapper-cross

Ericson2314 committed:
xbursttools: Fix eval at least so tests pass

Ericson2314 committed:
gcc-cross-wrapper: Remove in favor of newly-cross-supporting `cc-wrapper`

Ericson2314 committed:
cross stdenv adaptor: Support --host --build --target across the board

Packages get --host and --target by default, but can explicitly request
any subset to be passed as needed. See docs for more info.

rustc: Avoid hash breakage by using the old (ignored)
dontSetConfigureCross when not cross building

Ericson2314 committed:
cross stdenv adaptor: Remove `ccCross` `binutils` attrs and binutils extra buildDepends

It now has the correct wrapped tools and nothing else is needed.

Ericson2314 committed:
cc-wrapper: Remove `stdenv.is*` for `targetPlatform.is*`

Modify bootstrapping stdenvs to make sure `targetPlatform` also passed.

Ericson2314 committed:
cc-wrapper: Add support for no libc whatsoever

Ericson2314 committed:
cc-wrapper: Salt environment variables with LLVM triples

We now (on cross) require per-target flag interposition by putting the
triple in the names of the relevant environment variables, e.g:

export NIX_arm_unknown_linux_gnu_CFLAGS_COMPILE=...

The wrapper also has a `infixSalt` attribute (and "_" prefixed and
suffixed variants) to assist downstream packages.

Note how that the dashes are replaced to keep the identifier valid.
Using names like this allows us to keep the settings for different
compilers seperate.

I think it might be even better to use names like `NIX_{BUILD,HOST}...`
using the platform's role rather than the platform itself, but this
would be more work as the previous stages' tools would have to be re-
wrapped to take on their new role. I therefore didn't do this for now,
but that route should be thoroughly explored in the future.

Ericson2314 committed:
cc-wrapper: Simplify and correct logic to chose dynamic linker library

Ericson2314 committed:
cc-wrapper: Temporarily got to great lengths to avoid mass-rebuild

Eventually we should avoid this "pre-wrapping" and just update those
files in nixpkgs. This mass-rebuild change is best done along with
those needed to reduce the disparity between native and cross (i.e.
making native the "identity cross").

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -10,6 +10,10 @@
 , isGNU ? false, isClang ? cc.isClang or false, gnugrep ? null
+, # Prefix for binaries. Customarily ends with a dash separator.
+  prefix ? stdenv.lib.optionalString (targetPlatform != hostPlatform)
+                                     (targetPlatform.config + "-")
```
Should this be an argument for cc-wrapper? If you want a different prefix shouldn't you change the targetPlatform?

Ericson2314 replied:
It used to not be a parameter, but I figured technically prefixing is a matter of taste, and users might want things to not be prefixed.

But it probably is a bad idea is the underlying compiler will already be prefixed or not, and as you noticed the llvm stuff isn't prefixed so it's conditional too.

So either I can just let-bind it, or I'll have a separate "in prefix" and "out prefix", and define is prefix to be non-`""` only when cross compiling *and* using GCC. Your call :).

matthewbauer replied:
Ok it's fine as is if you can just override it when you need to

Ericson2314 replied:
I'll save the fancy stuff for later I think. let-bind for now!

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
+        ln -s ${prefix}gcc $out/bin/${prefix}cc
+        export real_cc=${prefix}gcc
+        export real_cxx=${prefix}g++
       elif [ -e $ccPath/clang ]; then
```
If I follow correctly, this should be ```${prefix}clang```.

Ericson2314 replied:
LLVM is multi-target---the unwrapped clang binaries support everything and thus don't need a prefix (which would require needless extra building).

Wrapped Clang however is target-specific as libc is target-specific, and even if libc is not part of cc-wrapper, we still want separate wrappers scrips so `NIX_CFLAGS_COMPILE` and friends don't get jumbled up (I sed those environment variables in a follow up commit).

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
+      if [ -e $ccPath/${prefix}g++ ]; then
+        wrap ${prefix}g++ ${./cc-wrapper.sh} $ccPath/${prefix}g++
+        ln -s ${prefix}g++ $out/bin/${prefix}c++
       elif [ -e $ccPath/clang++ ]; then
```
here too? unless something is special about clang

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
-        wrap clang++ ${./cc-wrapper.sh} $ccPath/clang++
-        ln -s clang++ $out/bin/c++
+        wrap ${prefix}clang++ ${./cc-wrapper.sh} $ccPath/clang++
+        ln -s ${prefix}clang++ $out/bin/${prefix}c++
```
Ok so I'm not sure if I like this whole prefix thing. We generally just assume that /bin/cc and /bin/c++ are the *compilers I want to use*. Why not just leave it as is, and specify the compiler at the derivation level? The prefix thing just seems redundant when applied to Nix. I know that historically "gcc/cc" has referred to the "native" compiler, but in the Nix context, it's the **stdenv* * compiler right? Maybe I am just misunderstanding the purpose of prefixes but I'm just thinking of how annoying it would be to have to specify the right compiler I want to use for each mkDerivation call. For MacOS support, we do a "substituteInPlace --replace gcc cc" to get the clang compiler to be supported. The assumption is we will always have a compiler named "cc" that is available no matter what.

Ericson2314 replied:
Some derivations need to build stuff to be run as part of the build. This seems odd but seems to be quite common, actually. For those ones, we need to have both compilers on the PATH and not have them clobber.

In some grand future where we turn all makesfiles into Nix or something, hopefully we can do away with this, but until then we're out of luck.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
       export default_cxx_stdlib_compile="${
-        if stdenv.isLinux && !(cc.isGNU or false)
-          then "-isystem $(echo -n ${cc.gcc}/include/c++/*) -isystem $(echo -n ${cc.gcc}/include/c++/*)/$(${cc.gcc}/bin/gcc -dumpmachine)"
+        if targetPlatform.isLinux && !(cc.isGNU or false)
```
Ah! but this is not the `or` function (which also exists!) but the `or` operator providing a fallback if `isGNU` isn't present.

[N.B. this `or` works analogously to the `?` in a function pattern (`{ x ? foo }: bar`). I want to make an RFC so both are deprecated and `??` is used instead. `?` in an expression is unrelated and would stay the same.]

matthewbauer reviewed (approved):
OK sounds good - at least it all makes some sense to me now!

Ericson2314 commented:
@Dridus Thanks again for agreeing to try shepherding this!

I've massively cleaned up the history here, and cherry-picked your zlib fix. Also I rewrote `dontSetConfigureCross` into something much nicer.

http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-scratch is now set up to do all our cross tests for this PR. It should periodically pull the underlying branch every so often, and schedule builds accordingly. I can bump up the priority of the builds too. 

Basically, the goal is to adding commits fixing packages (well, I hope the prior commits are OK!) until there are no regressions.

Ericson2314 commented:
FYI eval errors are because `meta.platforms` is now being checked against the host platform.

Ericson2314 committed:
podofo: Simplify derivation

Bunch of junk for linux and cross compilation that's no longer needed

Ericson2314 committed:
misc: Remove almost all uses of stdenv.{ccCross,binutils}

Ericson2314 committed:
dropbear: Remove now-unneeded crossAttrs

Commit:
zlib: don't really override CC at ay point

but leave the preConfigure script as it was when not cross building in order to avoid hash breakage

Commit:
cc-wrapper: when cross building, wedge in setting AR, AS, and LD env vars to their cross building versions

I think it's ok to export things which aren't wrapped. The cc-wrapper
can be thought of as responsible for all of binutils and the c
compiler, only wrapping those binaries which are necessary to
interposition---as opposed to all binaries it thinks are relevaant.

Conversely, adding the setup hook to the unwrapped compilers would be
unforunate as hooks are ugly hacks and the compilers themselves take
a long time to rebuild. Better to wholely separate "pure packages" from
hacks.

Ericson2314 committed:
vim: Modernize derivation, hopefully fixing cross

Ericson2314 committed:
bash: Modernize derivation, hopefully fixing cross compilation

Ericson2314 committed:
cc-wrapper: Make the cross compiler use a crossEnvHook

This is a hack until stdenv's setup.hs gets overhauled

Ericson2314 committed:
busybox: Modernize and fix cross

Ericson2314 committed:
ed: Modernize and fix for cross

Ericson2314 committed:
gnu patch: Fix for cross

Ericson2314 committed:
cross stdenv adaptor: Disable checkPhase by default

Before gcc-cross-wrapper did this unconditionally

Ericson2314 committed:
mpfs: Fix for cross by disabling check

check phase is now only disabled by *default* for cross, but mpfr
was forcing it always.

Ericson2314 committed:
libelf: Modernize and fix for cross

Ericson2314 committed:
binutils: Fix build != host == target

Ericson2314 committed:
libsigsegv: Fix for cross by disabling check

Check phase is now only disabled *by default* for cross, but this
package was forcing it unconditionally.

Ericson2314 committed:
pcre: Modernize and fix cross

Ericson2314 committed:
libpmc: Fix for cross by disabling check

Check phase is now only disabled *by default* for cross, but this
package was forcing it unconditionally.

Commit:
findutils: modernize and fix cross

Ericson2314 committed:
unzip: Fix for cross

Commit:
ddrescue: modernize and fix cross (hopefully)

Carefully do not add attributes for cross unless in cross, to avoid hash breakage

Ericson2314 committed:
lynx: Fix for cross

Needed C toolchain targeting build platform

Ericson2314 committed:
mpg123: Modernize and fix for cross

Do not even think about configureFlags unless in cross, to avoid hash
breaking when not in cross.

Commit:
gcc5: set --build and --host in the case where those differ and make sure to have a build cc-wrapper around as well

Ericson2314 commented:
So I hydra has all cross builds passing but the build != Host == target GCCs. But I have a feeling they were broken anyways. Could anyone test, say, the aarch64 one from master and confirm?

CC @grahamc :)

Commit:
podofo: unsimplify derivation a bit to avoid hash breakage

grahamc commented:
Tested by applying this diff:

```patch
diff --git a/pkgs/stdenv/linux/bootstrap-files/aarch64.nix b/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
index ca7e84502a..99d45d7f81 100644
--- a/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
+++ b/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
@@ -4,8 +4,6 @@
     sha256 = "12qcml1l67skpjhfjwy7gr10nc86gqcwjmz9ggp7knss8gq8pv7f";
     executable = true;
   };
-  bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://nixos-arm.dezgeg.me/bootstrap-aarch64-2017-03-11-bb3ef8/bootstrap-tools.tar.xz;
-    sha256 = "1075d5n4yclbhgisi6ba50601mw3fhivlkjs462qlnq8hh0xc7nq";
-  };
+  bootstrapTools = /nix/store/lwg73g7mxsbpv1nr2xhaam61b42rl596-stdenv-bootstrap-tools-cross-aarch64-unknown-linux-gnu/on-server/bootstrap-tools.tar.xz;
+  # bootstrapTools = "/nix/store/lwg73g7mxsbpv1nr2xhaam61b42rl596-stdenv-bootstrap-tools-cross-aarch64-unknown-linux-gnu/on-server/bootstrap-tools.tar.xz";
 }
```

and running `nix-build . -A hello` worked great:

```
$ file result/bin/hello 
result/bin/hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /nix/store/9br8761qwvvg7m301g47xz7yn9vm0ci7-glibc-2.25/lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, not stripped

$ ./result/bin/hello 
Hello, world!
```

Ericson2314 commented:
To be clear, that is *without* this PR. So the `gcc build != host == target` builds failures are indeed a regression.

Dridus commented on pkgs/build-support/cc-wrapper/default.nix:
this substitution doesn't seem to work on non-bash and older bashes (3.2)

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
Do we use those? I'm fine changing it to is awk or whatever.

LnL7 commented on pkgs/build-support/cc-wrapper/default.nix:
Both bash and `stdenv.shell` are 4.4 now AFAIK.

Dridus commented on pkgs/build-support/cc-wrapper/default.nix:
sorry, should have added context... it's true that in the build chain it'll be okay because all nix-controlled bash versions default to 4.4, but when using nix-shell it'll by default use the system bash, which on at least macOS 10.11 by default is 3.2. I can work around it with NIX_BUILD_SHELL=, but it makes for pretty awkward use of nix-shell when using the cross env.

dtzWill commented on pkgs/build-support/cc-wrapper/default.nix:
The wrappers' shebang should point to an appropriate bash to ensure things work as expected, I'd think.  Quick check says my local cc-wrapper's are this way, is that not happening on Darwin?

Dridus commented on pkgs/build-support/cc-wrapper/default.nix:
this is in the setup hook for the wrapper, not the wrapper proper. I imagine it wouldn't use the shebang anyway since it should be sourced into the enclosing shell environment? in any case, it's running in bash-3.2 on darwin

dtzWill commented on pkgs/build-support/cc-wrapper/default.nix:
ah, okay. awk it is, then! (or whichever)

Dridus commented on pkgs/build-support/cc-wrapper/default.nix:
I have a patch with `export $(echo "$CMD" | tr "[:lower:]" "[:upper:]")=${prefix}''${CMD}` that I'm waiting for a rebuild to confirm works correctly. this is on account of my non-knowledge of awk 😅

Ericson2314 committed:
cc-wrapper: Port over finding of dynamic linker from gcc-cross-wrapper

At the next break, we can deduplicate and go for the best of both
worlds.

Ericson2314 committed:
mingw-w64: Clean up, especially clarifying staging

Ericson2314 commented:
Lynx is failing because something requires wants dynamic linking?
```
configure:19231: checking for tgoto in -lncurses
configure:19258: armv6l-unknown-linux-gnueabihf-cc -o conftest -O2  -D_FILE_OFFSET_BITS=64  -D_GNU_SOURCE  -DLINUX  -I/nix/store/hhl0iwwy4r0c8iyg0xj0qslqgqzvd0g8-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf-dev/include  conftest.c -lncurses  -L/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib -lssl -lcrypto >&5
/nix/store/l28dg0fy9553i4cg9g41zs3f2wwcjnrb-armv6l-unknown-linux-gnueabihf-binutils-2.28/bin/armv6l-unknown-linux-gnueabihf-ld: warning: libdl.so.2, needed by /nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libssl.so, not found (try using -rpath or -rpath-link)
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dladdr@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlsym@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlerror@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlopen@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlclose@GLIBC_2.4'
collect2: error: ld returned 1 exit status
configure:19261: $? = 1
configure: failed program was:
#line 19239 "configure"
#include "confdefs.h"

/* Override any gcc2 internal prototype to avoid an error.  */
#ifdef __cplusplus
extern "C"
#endif
/* We use char because int might match the return type of a gcc2
   builtin and then its argument prototype would still apply.  */
char tgoto ();
int
main ()
{
tgoto ();
  ;
  return 0;
}
configure:19278: result: no
```

Ericson2314 commented:
Ok, I'm starting to think while there definitely is some regressions, the PR just is just getting bigger and older and it would be nice to merge and then deal with them---not many people are doing cross compilation anyways.

The known regressions are currently:

 - gcc `build != target == host`, only used when bootstrapping a new platform
 - mingw_w64, obsoleted by @DavidEGrayson's work long-term anyways
 - lynx on Raspberry Pi, just a random test to my knowledge

If we choose to deal with those later, the remain work is:
 - [ ] Use `tr` for old bash support (about to rebase for that)
 - [ ] Get a review as this is big

It's been a while since I had one of these  reviewed. Would @bjornfor @dezgeg be interested? @dezgeg I also give you final say on whether it's OK to temporarily break gcc like that.

bjornfor commented:
I looked at all the commits in this PR just now. Big +1 from me. (Awesome job!)

Disclaimer: I didn't test anything. I don't cross compile with Nix (yet).

Ericson2314 committed:
mingw-w64: Depend on own headers derivation

Without this, a `#include <float.h>` resolves incorrectly. Either the
headers weren't on the include path at all, or they only were for
local, not system, imports.

What's weird is this used to not be a problem. Not sure what other
change in e.g. cc-wrapper would affect this.

Ericson2314 commented:
@bjornfor thanks! And don't worry so much about actually building things. Hydra's got us covered for that.

Ericson2314 committed:
coreutils: Modernize and fix for cross

"--host=..." needs to be passed but crossAttrs overrides that.

Ericson2314 commented:
Fixed the darwin mass rebuild and applied @Dridus's `tr` patch.

Ericson2314 commented:
@dezgeg Any thoughts on the gcc regression? I'm very tempted to merge tomorrow so I can get on with the follow up.

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -45,6 +45,12 @@ let
   default_cxx_stdlib_compile=optionalString (targetPlatform.isLinux && !(cc.isGNU or false))
     "-isystem $(echo -n ${cc.gcc}/include/c++/*) -isystem $(echo -n ${cc.gcc}/include/c++/*)/$(${cc.gcc}/bin/gcc -dumpmachine)";
+
+  dashlessTarget = stdenv.lib.replaceStrings ["-"] ["_"] targetPlatform.config;
```
It looks like this kind of breaks with crossOverlays where targetPlatform == hostPlatform, but we want to only pass certain variables to cc-wrapper. For instance, this is used on macOS for semi-static binaries:

```
  # Best effort static binaries. Will still be linked to libSystem,
  # but more portable than Nix store binaries.
  makeStaticDarwin = stdenv: stdenv // {
    mkDerivation = args: stdenv.mkDerivation (args // {
      NIX_CFLAGS_LINK = toString (args.NIX_CFLAGS_LINK or "")
                      + " -static-libgcc";
      nativeBuildInputs = (args.nativeBuildInputs or []) ++ [ (makeSetupHook {
        substitutions = {
          libsystem = "${stdenv.cc.libc}/lib/libSystem.B.dylib";
        };
      } ../stdenv/darwin/portable-libsystem.sh) ];
    });
  };
```

NIX_CFLAGS_LINK should only be passed to nativeBuildInputs compilers. But, it is also being passed to depsBuildBuild compilers which are not guaranteed to be gcc. Can we choose another value for the salt? Not sure how to do this but it seems like every "cc-wrapper" should have a unique salt. Could we use the output path's hash?

matthewbauer replied:
Avoiding the salt sounds good! This ends up kind of polluting the environment anyway.

Ericson2314 replied:
Note to self: env hook roles prevent getting rid of salt. Either we make it so the wrapper scripts consume no variables, or we make propagation logic happen on the nix level.

matthewbauer replied:
This is mitigated by the fact that we can use llvm for cross now. I think we could just the same compiler for cross compilation as we do in stdenv.
